### ISSUE TITLE
Get namespaceability from OpenApi

### ISSUE THREAD

**Issue by monopole** [kubernetes-sigs/kustomize#2861]
To determine if a resource `Kind` (e.g. `Deployment`, `ClusterRole`, `Namespace`, etc.) is namespaceable or not, we currently use a hardcoded list.

https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/types.go#L102

We should instead try to get this information from openapi schema data.

**Comment by haiyanmeng** [comment]
@monopole , I cannot find the field in https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/openapi/kubernetesapi/swagger.json which can help determine the namespaceability of a resource kind. 
Am I missing sth?

**Comment by monopole** [comment]
I'm afraid this issue implies fixing it upstream first.

@apelisse might have time to comment here.

**Comment by apelisse** [comment]
Yeah, this is only available through discovery right now. I don't know how other people feel about it, but that seems important to me.

Since this information already exists (and is simply a field in CRDs for example), I wouldn't expect this to be a problem.

**Comment by apelisse** [comment]
If you really want to solve the problem quickly, an alternative/work-around is to take the full openapi from kubernetes (https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/swagger.json) which contains the list of operations. Then the operations have a path, which include the `{namespace}` sort of template marker in the path, e.g.:
```
    "/apis/settings.k8s.io/v1alpha1/namespaces/{namespace}/podpresets/{name}": {
```

Which also has this description for the path parameter:
```
        {
          "description": "object name and auth scope, such as for teams and projects",
          "in": "path",
          "name": "namespace",
          "required": true,
          "type": "string",
          "uniqueItems": true
        },
```

**Comment by mortent** [comment]
I think we can also consider adding this to the openapi schema that we ship with kyaml by using an extension. We can add this information to the schema by using the information from the discovery client. We should add this functionality as part of kyaml, so we can use it every time we need to update the schema.
If this information is included in upstream, then we can migrate to that solution.

**Comment by monopole** [comment]
+1

**Comment by mortent** [comment]
And if we really want to go all out on this, see if we can include the [resource graph](https://github.com/kubernetes-sigs/kustomize/blob/5b394b207973146e6f5c940c389470b5fdafd2be/api/konfig/builtinpluginconsts/namereference.go) information into the openapi schema as well.

**Comment by apelisse** [comment]
If you decide to add this specifically to the openapi (even if only in kyaml), please let me know so we can maybe find a way to make it compatible with what we would do for kubernetes.

**Comment by apelisse** [comment]
> I think we can also consider adding this to the openapi schema that we ship with kyaml by using an extension. We can add this information to the schema by using the information from the discovery client. We should add this functionality as part of kyaml, so we can use it every time we need to update the schema.
If this information is included in upstream, then we can migrate to that solution.

Why is this better than what I have suggested?

**Comment by mortent** [comment]
> Why is this better than what I have suggested?

I'm not saying it is better. It is just another way we could do something like this. I think the key here is that we probably want to do something sooner than what is possible if we have to wait for upstream. If we can infer the namespaceability from what is already in the openapi schema, that sounds easier than using the discovery client.

**Comment by apelisse** [comment]
Let me try to look at the pros and cons a little closer.

My [suggestion](https://github.com/kubernetes-sigs/kustomize/issues/2861#issuecomment-676833212) requires that you import the whole openapi (definitions and operations), which is going to be bigger. On the other hand, it means that there is no preprocessing of the openapi to do (and it might work with openapi coming from other places). It adds more work at runtime since you have to look at operations to decide if a given type is namespaced or not.

Your [suggestion](https://github.com/kubernetes-sigs/kustomize/issues/2861#issuecomment-680346216) will require more preparation work for the openapi (make it less portable), but will probably be simpler at runtime since the information will be readily available.

One of my concern too is that there is not a 1-to-1 mapping between kinds and resources, and I'm curious if some kind can be namespaced for some resources and sometimes not. I don't think that's a big concern though.

**Comment by Shell32-Natsu** [comment]
kustomize should use the information from the openapi and remove the hardcoded list.

/assign

**Comment by Shell32-Natsu** [comment]
/unassign

@natasha41575 will work on this.

**Comment by monopole** [comment]
Fixed by #3136
